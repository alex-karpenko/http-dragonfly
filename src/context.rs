use hyper::{http::request::Parts, Body, Response};
use once_cell::sync::OnceCell;
use regex::Regex;
use std::collections::HashMap;
use std::env;
use std::net::SocketAddr;

use crate::config::target::TargetConfig;

const CTX_APP_NAME: &str = env!("CARGO_PKG_NAME");
const CTX_APP_VERSION: &str = env!("CARGO_PKG_VERSION");

pub type ContextMap = HashMap<String, String>;

#[derive(Debug)]
pub struct Context<'a> {
    own: ContextMap,
    parent: Option<&'a Context<'a>>,
}

impl<'a> Context<'a> {
    pub fn root(root_env: impl RootEnvironment) -> &'static Context<'a> {
        static ROOT_CONTEXT: OnceCell<Context> = OnceCell::new();

        let ctx = ROOT_CONTEXT.get_or_init(|| {
            let mut ctx = ContextMap::new();

            ctx.insert("CTX_APP_NAME".into(), CTX_APP_NAME.into());
            ctx.insert("CTX_APP_VERSION".into(), CTX_APP_VERSION.into());
            ctx.extend(root_env.get_environment());

            Context {
                own: ctx,
                parent: None,
            }
        });

        ctx
    }

    pub fn with(&self, own: ContextMap) -> Context {
        Context {
            own,
            parent: Some(self),
        }
    }

    pub fn get(&self, var: &String) -> Option<&String> {
        // Try own context
        if let Some(value) = self.own.get(var) {
            Some(value)
            // or get parent if it's
        } else if let Some(parent) = &self.parent {
            // and try parent
            if let Some(value) = parent.get(var) {
                Some(value)
            } else {
                None
            }
        } else {
            None
        }
    }

    pub fn iter(&'a self) -> Box<dyn Iterator<Item = (&'a String, &'a String)> + 'a> {
        let iter = Box::new(self.own.iter());
        Box::new(Iter {
            ctx: self,
            iter,
            finished: false,
        })
    }

    pub fn with_request(
        &'a self,
        addr: &'a SocketAddr,
        req: &'a Parts,
        listener_name: String,
    ) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_LISTENER_NAME
        // CTX_REQUEST_SOURCE_IP
        // CTX_REQUEST_METHOD
        // CTX_REQUEST_HOST
        // CTX_REQUEST_PATH
        // CTX_REQUEST_QUERY
        own.insert("CTX_LISTENER_NAME".into(), listener_name);
        own.insert("CTX_REQUEST_SOURCE_IP".into(), addr.ip().to_string());
        own.insert("CTX_REQUEST_METHOD".into(), req.method.to_string());
        own.insert("CTX_REQUEST_PATH".into(), req.uri.path().to_string());
        if let Some(host) = req.uri.host() {
            own.insert("CTX_REQUEST_HOST".into(), host.to_lowercase());
        }
        if let Some(query) = req.uri.query() {
            own.insert("CTX_REQUEST_QUERY".into(), query.to_lowercase());
        }

        // CTX_REQUEST_HEADERS_<UPPERCASE_HEADER_NAME>
        req.headers.iter().for_each(|(n, v)| {
            let n = n.as_str().to_uppercase().replace('-', "_");
            let v = v.to_str().unwrap_or("").to_string();
            own.insert(format!("CTX_REQUEST_HEADERS_{n}"), v);
        });

        self.with(own)
    }

    pub fn with_target(&'a self, cfg: &'a TargetConfig) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_TARGET_ID
        // CTX_TARGET_HOST
        own.insert("CTX_TARGET_ID".into(), cfg.id());
        own.insert("CTX_TARGET_HOST".into(), cfg.host());

        self.with(own)
    }

    pub fn with_response(&'a self, resp: &Response<Body>) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_RESPONSE_HEADERS_<UPPERCASE_HEADER_NAME>
        // CTX_RESPONSE_STATUS
        own.insert("CTX_RESPONSE_STATUS".into(), resp.status().to_string());
        resp.headers().iter().for_each(|(n, v)| {
            let n = n.as_str().to_uppercase().replace('-', "_");
            let v = v.to_str().unwrap_or("").to_string();
            own.insert(format!("CTX_RESPONSE_HEADERS_{n}"), v);
        });

        self.with(own)
    }
}

pub struct Iter<'a> {
    ctx: &'a Context<'a>,
    iter: Box<dyn Iterator<Item = (&'a String, &'a String)> + 'a>,
    finished: bool,
}

impl<'a> Iterator for Iter<'a> {
    type Item = (&'a String, &'a String);

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(next) = self.iter.next() {
            Some(next)
        } else if self.ctx.parent.is_none() {
            None
        } else if !self.finished {
            self.finished = true;
            self.iter = self.ctx.parent.unwrap().iter();
            self.iter.next()
        } else {
            None
        }
    }
}

pub trait RootEnvironment {
    fn get_environment(&self) -> ContextMap;
}

pub struct RootOsEnvironment<'a> {
    env_mask_regex: &'a str,
}

impl<'a> RootEnvironment for RootOsEnvironment<'a> {
    fn get_environment(&self) -> ContextMap {
        let mut env_ctx = ContextMap::new();
        let re = Regex::new(self.env_mask_regex).expect("invalid environment filter regex");
        env::vars().for_each(|(k, v)| {
            if re.is_match(&k) {
                env_ctx.insert(k, v);
            }
        });

        env_ctx
    }
}

impl<'a> RootOsEnvironment<'a> {
    pub fn new(env_mask_regex: &'a str) -> Self {
        Self { env_mask_regex }
    }
}
