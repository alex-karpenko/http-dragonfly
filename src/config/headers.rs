use crate::context::Context;
use hyper::{
    header::{HeaderName, HeaderValue},
    HeaderMap,
};
use serde::{
    de::{self, MapAccess, Visitor},
    Deserialize, Deserializer, Serialize,
};
use shellexpand::env_with_context_no_errors;
use tracing::debug;

pub type HeadersTransformsList = Vec<HeaderTransform>;

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct HeaderTransform {
    action: HeaderTransformActon,
    value: Option<String>,
}

impl HeaderTransform {
    fn action(&self) -> &HeaderTransformActon {
        &self.action
    }

    fn value(&self) -> Option<&String> {
        self.value.as_ref()
    }
}

#[derive(Deserialize, Debug, Clone, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "lowercase")]
pub enum HeaderTransformActon {
    Add(String),
    Update(String),
    Drop(String),
}

impl<'de> Deserialize<'de> for HeaderTransform {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize, Debug)]
        #[serde(deny_unknown_fields, rename_all = "lowercase")]
        enum Fields {
            Drop,
            Add,
            Update,
            Value,
        }

        struct HeaderTransformVisitor;
        impl<'de> Visitor<'de> for HeaderTransformVisitor {
            type Value = HeaderTransform;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("struct HeaderTransform")
            }

            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
            where
                A: MapAccess<'de>,
            {
                let mut action: Option<HeaderTransformActon> = None;
                let mut value: Option<String> = None;

                // Extract all fields
                while let Some(key) = map.next_key()? {
                    match key {
                        Fields::Add => {
                            if action.is_some() {
                                return Err(de::Error::duplicate_field("add"));
                            }
                            action = Some(HeaderTransformActon::Add(map.next_value::<String>()?))
                        }
                        Fields::Drop => {
                            if action.is_some() {
                                return Err(de::Error::duplicate_field("drop"));
                            }
                            action = Some(HeaderTransformActon::Drop(map.next_value::<String>()?))
                        }
                        Fields::Update => {
                            if action.is_some() {
                                return Err(de::Error::duplicate_field("update"));
                            }
                            action = Some(HeaderTransformActon::Update(map.next_value::<String>()?))
                        }
                        Fields::Value => {
                            if value.is_some() {
                                return Err(de::Error::duplicate_field("value"));
                            }
                            value = Some(map.next_value::<String>()?)
                        }
                    }
                }

                if let Some(action) = action {
                    match action {
                        HeaderTransformActon::Add(_) | HeaderTransformActon::Update(_) => {
                            if value.is_none() {
                                return Err(de::Error::missing_field("value"));
                            }
                        }
                        HeaderTransformActon::Drop(_) => {
                            if value.is_some() {
                                return Err(de::Error::custom(
                                    "unknown field `value` in action drop",
                                ));
                            }
                        }
                    }
                    Ok(HeaderTransform { action, value })
                } else {
                    Err(de::Error::missing_field(
                        "action should be one of add/drop/update",
                    ))
                }
            }
        }

        const FIELDS: &[&str] = &["add", "drop", "update", "value"];
        deserializer.deserialize_struct("HeaderAction", FIELDS, HeaderTransformVisitor)
    }
}

pub trait HeadersTransformator<'a> {
    fn transform(&'a self, headers: &'a mut HeaderMap, ctx: &Context);
}

impl<'a> HeadersTransformator<'a> for HeadersTransformsList {
    fn transform(&'a self, headers: &'a mut HeaderMap, ctx: &Context) {
        for transform in self {
            match transform.action().clone() {
                HeaderTransformActon::Add(key) => {
                    if !headers.contains_key(key.clone()) {
                        let value = transform.value().as_ref().unwrap().as_str();
                        let value = env_with_context_no_errors(value, |v| ctx.get(&v.into()));
                        let key = HeaderName::from_bytes(key.as_bytes()).unwrap();
                        debug!("add: name={key}, value={value}");
                        headers.insert(&key, HeaderValue::from_str(&value).unwrap());
                    }
                }
                HeaderTransformActon::Update(key) => {
                    if headers.contains_key(&key) {
                        let value = transform.value().as_ref().unwrap().as_str();
                        let value = env_with_context_no_errors(value, |v| ctx.get(&v.into()));
                        let debug_key = key.clone();
                        let key = HeaderName::from_bytes(key.as_bytes()).unwrap();
                        let old = headers.insert(key, HeaderValue::from_str(&value).unwrap());
                        if let Some(old) = old {
                            debug!(
                                "update: name={}, old={}, new={}",
                                debug_key,
                                old.to_str().unwrap(),
                                value
                            );
                        } else {
                            debug!("update: name={}, old=, new={}", debug_key, value);
                        }
                    }
                }
                HeaderTransformActon::Drop(key) => {
                    if key == "*" {
                        debug!("drop: all headers");
                        headers.clear();
                    } else {
                        let debug_key = key.clone();
                        let old = headers.remove(key);
                        if let Some(old) = old {
                            debug!("drop: name={}, old={}", debug_key, old.to_str().unwrap());
                        } else {
                            debug!("drop: name={}, old=", debug_key);
                        }
                    }
                }
            };
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, assert_ron_snapshot};

    use crate::context::test_context::get_test_ctx;

    use super::*;

    #[test]
    fn deserialize_ok_header_transform() {
        let add_action: HeaderTransform =
            serde_json::from_str(r#"{"add": "new_header", "value": "new_value"}"#).unwrap();
        assert_ron_snapshot!(add_action);

        let update_action: HeaderTransform =
            serde_json::from_str(r#"{"update": "new_header", "value": "new_value"}"#).unwrap();
        assert_ron_snapshot!(update_action);

        let drop_action: HeaderTransform =
            serde_json::from_str(r#"{"drop": "old_header"}"#).unwrap();
        assert_ron_snapshot!(drop_action);
    }

    #[test]
    fn deserialize_wrong_header_transform() {
        let wrong_json = [
            r#"{"wrong_action": "new_header", "value": "new_value"}"#,
            r#"{"wrong_action": "new_header"}"#,
            r#"{"add": "new_header", "wrong_value": "new_value"}"#,
            r#"{"add": "new_header", "value": 1}"#,
            r#"{"add": "new_header"}"#,
            r#"{"update": "new_header", "wrong_value": "new_value"}"#,
            r#"{"update": "new_header", "value": 1}"#,
            r#"{"update": "new_header"}"#,
            r#"{"drop": "old_header", "wrong_value": "old_value"}"#,
            r#"{"drop": "old_header", "value": "old_value"}"#,
            r#"{"drop": "old_header", "value": 1}"#,
        ];

        for test_item in wrong_json {
            let result: Result<HeaderTransform, serde_json::Error> =
                serde_json::from_str(test_item);
            assert!(
                result.is_err(),
                "unexpected deserialization of `{}`",
                test_item
            );
        }
    }

    #[test]
    fn transforms() {
        let ctx = get_test_ctx();
        assert_ron_snapshot!(ctx, {".own" => insta::sorted_redaction(), ".parent.own" => insta::sorted_redaction()});

        let transforms: HeadersTransformsList = serde_json::from_str(
            r#"
        [
            {"add": "X-Some-New-Header", "value": "good"},
            {"update": "X-Existing-Header", "value": "good"},
            {"drop": "X-Header-To-Drop"},
            {"add": "X-Existing-Header", "value": "wrong"},
            {"update": "X-Non-Exiting-Header", "value": "wrong"},
            {"add": "X-Env-Header", "value": "${TEST_ENV_HEADER_TO_ADD}"},
            {"update": "X-Env-Header-2", "value": "${TEST_ENV_HEADER_TO_ADD}"}
        ]
        "#,
        )
        .unwrap();
        assert_debug_snapshot!(transforms);

        let mut headers = HeaderMap::new();
        headers.insert("X-Header-To-Drop", "wrong".parse().unwrap());
        headers.insert("X-Existing-Header", "wrong".parse().unwrap());
        headers.insert("X-Header-Without-Changes", "good".parse().unwrap());
        headers.insert("X-Env-Header-2", "wrong".parse().unwrap());
        // Before transformation
        assert_debug_snapshot!(headers);

        transforms.transform(&mut headers, ctx);
        // After transformation
        assert_debug_snapshot!(headers);
    }
}
