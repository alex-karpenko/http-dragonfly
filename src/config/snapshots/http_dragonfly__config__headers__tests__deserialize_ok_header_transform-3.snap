---
source: src/config/headers.rs
expression: drop_action
---
HeaderTransform(
  action: drop("old_header"),
  value: None,
)
